### PR TITLE
[Docs] ELI5 Tutorial For Invocations

### PR DESCRIPTION
**Pull request by blessedcoolant**
Work in progress. Just putting it out there to see if its worth pursuing writing indepth detailed guides.

Can be previewed here: https://github.com/blessedcoolant/InvokeAI/blob/docs-invocations/docs/contributing/INVOCATIONS.md

### CODE REVIEW

**Comment by hipsterusername** [comment]
This type of documentation is going to be critical. Thanks for putting it together. I've given it a once over, but will get perspective of @brandonrising next week on how clear it is to fresh eyes, and where we can expand.

File: docs/contributing/INVOCATIONS.md [hunk 1]
```diff
@@ -1,8 +1,521 @@
...
+    type: Literal['resize'] = 'resize'
+
+    # Inputs
+    image: Union[ImageField, None] = Field(description="The input image", default=None)
+    width: int = Field(default=512, ge=64, le=2048, description="Width of the new image")
+    height: int = Field(default=512, ge=64, le=2048, description="Height of the new image")
+```
+
+As you might have noticed, we added two new parameters to the field type for
+`width` and `height` called `gt` and `le`. These basically stand for _greater
```
Comment (brandonrising) [inline docs/contributing/INVOCATIONS.md:125]: Says `gt` and `le` here, but in the code snippet above, height and width have `ge` instead of `gt`

Comment (blessedcoolant) [reply docs/contributing/INVOCATIONS.md:125]: Ah. Thanks for spotting that. I'll fix it up.

File: docs/contributing/INVOCATIONS.md [hunk 2]
```diff
@@ -1,8 +1,521 @@
...
+            image=ImageField(
+                image_name=output_image.image_name,
+                image_origin=output_image.image_origin,
+            ),
+            width=output_image.width,
+            height=output_image.height,
+        )
+```
+
+**Note:** Do not be overwhelmed by the `ImageOutput` process. InvokeAI has a
```
Comment (brandonrising) [inline docs/contributing/INVOCATIONS.md:251]: Will be great one day to add a doc reference here that explains all the services available when writing invocations (image.create/images.get_pil_image/etc). For now, maybe we could add a note pointing to [invocation_services.py](https://github.com/invoke-ai/InvokeAI/blob/main/invokeai/app/services/invocation_services.py) as a starting point for available utilities while working on invocations?

File: docs/contributing/INVOCATIONS.md [hunk 3]
```diff
@@ -1,8 +1,521 @@
...
+
+Let us create one as an example. Let us say we want to create a color input
+field that represents a color code. But before we start on that here are some
+general good practices to keep in mind.
+
+**Good Practices**
+
+- There is no naming convention for input fields but we highly recommend that
+  you name it something appropriate like `ColorField`.
+- It is not mandatory but it is heavily recommended to add a relevant
```
Comment (brandonrising) [inline docs/contributing/INVOCATIONS.md:292]: Perhaps we should make this mandatory haha

File: docs/contributing/INVOCATIONS.md [hunk 4]
```diff
@@ -1,8 +1,521 @@
...
+  `docstring` to describe your input field.
+- Keep your field in the same file as the Invocation that it is made for or in
+  another file where it is relevant.
+
+All input types a class that derive from the `BaseModel` type from `pydantic`.
```
Comment (brandonrising) [inline docs/contributing/INVOCATIONS.md:297]:
```suggestion
All input types are classes that derive from the `BaseModel` type from `pydantic`.
```

File: docs/contributing/INVOCATIONS.md [hunk 5]
```diff
@@ -1,8 +1,521 @@
...
+```python
+from pydantic import BaseModel
+
+class ColorField(BaseModel):
+    '''A field that holds the rgba values of a color'''
+    pass
+```
+
+Perfect. Now let us create our custom inputs for our field. This is exactly
+similar how you created input fields for your Invocation. All the same rules
```
Comment (brandonrising) [inline docs/contributing/INVOCATIONS.md:309]:
```suggestion
the same as how you created input fields for your Invocation. All the same rules
```

File: docs/contributing/INVOCATIONS.md (hunk 2 above)
Comment (blessedcoolant) [reply docs/contributing/INVOCATIONS.md:251]: Agreed. We can maintain a doc with available services and some small snippets on how to use them. Once we have the model manager finished and merged in, we should have more details on how to frame that.

File: docs/contributing/INVOCATIONS.md (hunk 3 above)
Comment (blessedcoolant) [reply docs/contributing/INVOCATIONS.md:292]: I was thinking about it too. The only mandatory thing right now is `Invocation` coz we use that in the UI to actually parse the name of the node but maybe it wont hurt to enforce this. We might not have any code related validators for it but because every node has to go through a review anyway, we could probably enforce it.

@psychedelicious Thoughts? Should we make devs use `Field` and `Output` to describe the respective types and fields?

**Comment by aluhrs13** [comment]
This looks great! Definitely valuable and worth investing in. One question for the advanced section probably, are invocations allowed to use `InvokeAIAppConfig`? Any best practices there?

**Comment by aluhrs13** [comment]
`self.is_intermediate` is used but never defined

**Comment by blessedcoolant** [comment]
> `self.is_intermediate` is used but never defined

Yes. I'll update this when I'm refining the PR later. I'm just waiting on all the services to settle in. `self.is_intermediate` is used to determine an intermediate image. Not applicable in this case.

> are invocations allowed to use InvokeAIAppConfig

Basically we'll want the dev to access the config via the service rather than trying to instantiate their own class of `InvokeAIAppConfig` .. so it should be `context.services.configuration.xxxx` to access any of the data provided by the app config.

**Comment by hipsterusername** [comment]
@blessedcoolant - Thoughts on the status/timing of this doc? Have it on the list of things to have ready for 3.0 launch

**Comment by blessedcoolant** [comment]
The docs are good for the backend stuff. They front end stuff is still WIP. But that's a big write up. So don't really have an ETA on that.

**Comment by hipsterusername** [comment]
> The docs are good for the backend stuff. They front end stuff is still WIP. But that's a big write up. So don't really have an ETA on that.

Ok - should we merge this in then? Or can merge over to the doc-updates branch if you want
